demo.py

```diff
@@ -25,115 +27,274 @@
+
+    # Constraint 3: Max distance to other new charging locations
+    if num_new_cs > 1:
+        for i in range(len(potential_new_cs_nodes)):
+            for j in range(i+1, len(potential_new_cs_nodes)):
+                ai = potential_new_cs_nodes[i]
+                aj = potential_new_cs_nodes[j]
+                dist = (-1*ai[0]**2 + 2*ai[0]*aj[0] - aj[0]**2 - ai[1]**2 
+                        + 2*ai[1]*aj[1] - aj[1]**2)
+                bqm.add_interaction(i, j, dist * gamma3)
+
+    # Constraint 4: Choose exactly num_new_cs new charging locations
+    bqm.update(dimod.generators.combinations(bqm.variables, num_new_cs, strength=gamma4))
+
+    return bqm
+
+def run_bqm_and_collection_solns(bqm, sampler, potential_new_cs_nodes):
+    """ Solve the bqm with the provided sampler to find new charger locations. """
+
+    sampleset = sampler.sample(bqm)
+
+    ss = sampleset.first.sample
+    new_charging_nodes = [potential_new_cs_nodes[k] for k, v in ss.items() if v == 1]
+
+    return new_charging_nodes
+
+def compute_soln_stats(pois, num_poi, charging_stations, num_cs, new_charging_nodes, num_new_cs):
+    """ Compute statistics on result scenario. """
+
+    # Compute average distance from new chargers to POIs
+    poi_ave_dist = 0
+    if num_poi > 0:
+        poi_ave_dist = [0 for _ in range(num_new_cs)]
+        for loc in pois:
+            for i in range(num_new_cs):
+                poi_ave_dist[i] += (1/num_poi)*(abs(new_charging_nodes[i][0]-loc[0])+abs(new_charging_nodes[i][1]-loc[1]))
+
+    # Compute average distance from new chargers to old chargers
+    old_cs_ave_dist = 0
+    if num_cs > 0:
+        old_cs_ave_dist = [0 for _ in range(num_new_cs)]
+        for loc in charging_stations:
+            for i in range(num_new_cs):
+                old_cs_ave_dist[i] += (1/num_cs)*(abs(new_charging_nodes[i][0]-loc[0])+abs(new_charging_nodes[i][1]-loc[1]))
+
+    # Compute average distance between new chargers
+    new_cs_dist = 0
+    if num_new_cs > 1:
+        new_cs_dist = 0
+        for i in range(num_new_cs):
+            for j in range(i+1, num_new_cs):
+                new_cs_dist += abs(new_charging_nodes[i][0]-new_charging_nodes[j][0])+abs(new_charging_nodes[i][1]-new_charging_nodes[j][1])
+
+    return poi_ave_dist, old_cs_ave_dist, new_cs_dist
+
+def printout_solution_to_cmdline(num_poi, new_charging_nodes, num_new_cs, num_cs, poi_ave_dist, old_cs_ave_dist, new_cs_dist):
+    """ Print solution statistics to command line. """
+
+    print("\nSolution returned: \n------------------")
+    
+    print("\nNew charging locations:\t\t\t\t", new_charging_nodes)
+
+    if num_poi > 0:
+        print("Average distance to POIs:\t\t\t", poi_ave_dist)
+
+    if num_cs > 0:
+        print("Average distance to old charging stations:\t", old_cs_ave_dist)
+
+    if num_new_cs > 1:
+        print("Distance between new chargers:\t\t\t", new_cs_dist)
+
+def build_output_image(G, pois, charging_stations, new_charging_nodes):
```

Comment:
This function does more than just building image. From the name, I'd expect it returns `plt` ready for displaying or saving to file. (That would allow output image name customization from the main program.)

Perhaps `save_output_map`?